Feature/92 create reusable Input components

kanta1207 opened the pull request:
## Overview

1. Implemented "Input" component
2. Based on it, implemented "SearchInput" component and "NumberInput" component


## Review Points
- I'll write a comment on each parts where I especially want you to take a look at. 

## Note
- Decided to create SearchInput component separately from Input component, because to add Search Icon, we needed to wrap input tag with div tag, which is unnecessary for normal inputs. Didn't want to make Input component too complex.

- Decided to create NumberInput component separately from Input component, because to default input with type "number" accept "e","E","+","-",and leading zero. We needed some logic to except these inputs when we want, and this logic will be unnecessary for normal inputs. Didn't want to make Input component too complex.

## Screenshots

### Input(Rounded), search input, number input
https://github.com/genesis-tech-tribe/nishiki-frontend/assets/99339182/53dcb512-0f79-4f58-9350-6604f4af584a

### Input(Square)
https://github.com/genesis-tech-tribe/nishiki-frontend/assets/99339182/501af181-419a-4bda-be92-6bbd0104ba4a

kanta1207 commented on src/components/ui/Icon.tsx:
```diff
@@ -44,8 +44,10 @@ interface IIconProps extends VariantProps<typeof iconVariants> {
   return <IconComponent className={cn(iconVariants({ color, size, className }))} />;
 };
+
+export { Icon, iconVariants };
```
export iconVariants to use the type in search input component

kanta1207 commented on tailwind.config.ts:
```diff
@@ -23,6 +23,7 @@ module.exports = {
       borderRadius: {
         xs: '0.125rem', // 2px
         sm: '0.25rem', // 4px
+        full: '9999px',
```
In design, 50px was the largest radius for now, but 9999px is better one for oval shaped UI components.
In default tailwind, 9999px is named as "full" .

kanta1207 replied:
Yeah you're right, we didn't need to explicitly define rounded-full...
I'll remove it!

kanta1207 commented on src/components/ui/Label.tsx:
```diff
@@ -5,6 +5,8 @@
  * Radix UI: https://www.radix-ui.com/primitives/docs/components/label
  */
+
+'use client';
```
Since it's using radix UI, Label component causes error without "use client"

kanta1207 replied:
I updated InputUsageExample.ts, by adding "Label" component in there.
Could you try render InputUsageExample in your local env and check if you get this error or not, please?
In my env it does cause this error, and "use client" in Label component will resolve the error 
<img width="339" alt="Screenshot 2023-12-16 at 19 48 10" src="https://github.com/genesis-tech-tribe/nishiki-frontend/assets/99339182/df4f7493-5f3f-43c0-be12-203299cb1d50">

uskayyyyy replied:
Thank you for checking it. It happened in my environment, too. It appears that the primitive label component utilizes the `onMouseDown` event internally.

Here is the source:
https://github.com/radix-ui/primitives/blob/main/packages/react/label/src/Label.tsx

kanta1207 replied:
Thank you for sharing the source! I should have put it in the first place... sorry! 
So I'll leave the "use client" as it is.

kanta1207 commented on src/components/ui/Input/Input.tsx:
```diff
@@ -0,0 +1,48 @@
+      variant: {
+        // naming of each variant should be aligned with the design team
+        rounded: 'rounded-full bg-white border border-gray px-6 py-4 placeholder:text-gray',
+        muted: 'border-none placeholder:text-gray',
```
muted is basically minimum design without any padding. Used in SearchInput component

kanta1207 replied:
It's already used in SearchInput, and except for that, I'm actually not sure where we're gonna use it.
Here's why I created "muted" variant. Please read through them and tell me your point of view 👍 

1.  Even if it will be only used in SearchInput, leaving "muted" variant doesn't have any negative effect to this Input component
2. It might be better than create "muted" input component separately or define it in inside of "SearchInput" component since it shares base styling classes.

uskayyyyy replied:
Oh, I'm sorry. I forgot to delete the comment. I wrote the comment before noticing it was used in the `SearchInput`.
I think you can leave it as a variant.

What I noticed it that `muted` should have a transparent background. When I tried it, it had a white background.

kanta1207 replied:
Nono, it was a good call out to consider tho.
I'll leave the variant as it is, and yeah,  transparent background would be definetely better! I'll set it!

kanta1207 commented on src/components/ui/Input/NumberInput.tsx:
```diff
@@ -0,0 +1,84 @@
+const e = ['e', 'E'];
+const dot = '.';
+
+const handleExceptionSymbols = (
```
The code for this logic might be a bit unnecessaly lengthy... If you come up with better one, please tell me 🙏

uskayyyyy replied:
Shall we have a conversation about this?
I am not even sure what this file is trying to achieve.

kanta1207 commented on src/components/ui/Input/NumberInput.tsx:
```diff
@@ -0,0 +1,84 @@
+export const NumberInput: FC<INumberInputProps> = ({
+  className,
+  variant = defaultVariant,
+  onKeyDown = handleExceptionSymbols,
```
You can override onKeyDown function in case you want

kanta1207 commented on src/styles/globals.css:
```diff
@@ -1,3 +1,9 @@
 @tailwind utilities;
+
+@layer base {
+  input[type='search']::-webkit-search-cancel-button {
```
This line was needed to remove input type "search"'s unnecessary cancel button

uskayyyyy replied:
I asked ChatGPT about this, and it suggested the following code:
```css
input[type='search']::-webkit-search-cancel-button {
    -webkit-appearance: none;
    appearance: none;
}
```

And I asked the difference between using `appearance: none` and `display: none`, and got the answer below:
> `appearance: none`: This property is used to remove default styling set by the browser. When applied to the `::-webkit-search-cancel-button` pseudo-element, it removes the default styling of the clear button in WebKit-based browsers (like Chrome and Safari). This is often the preferred method for removing the clear button because it specifically targets the element's default appearance without affecting its layout or functionality.

> `display: none`: This property removes an element from the document flow, meaning the element will not occupy any space in the layout. It effectively makes the element and its content invisible and non-interactive. While `display: none` can hide the clear button, it's a more forceful approach. It not only hides the button but also removes its space, which could potentially affect the layout or behavior of the input field.

kanta1207 replied:
Thank you for searching it!  Indeed, using `apperance : none` seems better! I'll fix it

uskayyyyy commented on src/components/ui/Input/InputUsageExample.tsx:
```diff
@@ -0,0 +1,19 @@
+/**
+ * This file is used as an example for the Dialog component.
```
Not Dialog

uskayyyyy commented on src/components/ui/Input/InputUsageExample.tsx:
```diff
@@ -0,0 +1,19 @@
+  return (
+    <div className="grid w-full max-w-sm items-center gap-1.5">
+      <Input type="text" id="groupName" placeholder="Group name" />
+      <Input variant={'square'} type="text" id="groupName" />
```
`variant="square"`

uskayyyyy commented on src/components/ui/Input/InputUsageExample.tsx:
```diff
@@ -0,0 +1,19 @@
+ * Once you're done with the example, you can delete this file.
+ */
+
+import { Input, NumberInput, SearchInput } from '..';
```
Using a relative path is not a good practice except `./`. You could use `./` for this case.

uskayyyyy commented on src/components/ui/Input/SearchInput.tsx:
```diff
@@ -0,0 +1,48 @@
+import { VariantProps } from 'class-variance-authority';
+import { FC, InputHTMLAttributes, SVGAttributes } from 'react';
+
+import { Icon, iconVariants } from '..';
```
Using a relative path is not a good practice except `./`.

uskayyyyy commented on src/components/ui/Input/SearchInput.tsx:
```diff
@@ -0,0 +1,48 @@
+      <Icon icon={icon} color={iconColor} size={iconSize} className="mr-2" />
+      <Input
+        type={type}
+        variant={'muted'}
```
`variant="muted"`

uskayyyyy replied:
variant="muted"

uskayyyyy commented on src/components/ui/Input/SearchInput.tsx:
Clicking somewhere close to the edge does not focus on the input. It is because the `padding` is used to define the size of the input tag in `Input`. Ideally, clicking wherever inside the border should focus the input.

<img width="400" alt="image" src="https://github.com/genesis-tech-tribe/nishiki-frontend/assets/99148565/05d029d1-6bad-4ea0-80e3-afe90c547679">

<br><br>

Additionally, if there are no multiple variations of the search field in Figma, accepting an `icon` component and the input `type` might be redundant.

kanta1207 replied:
You're definetely right about the padding. I'll fix it, thank you for pointing it out!

`Additionally, if there are no multiple variations of the search field in Figma, accepting an icon component and the input type might be redundant.`

About this, I tried to leave flexibility for user. But indeed there's no multiple variations of the search field for now, I'll remove that props maybe to simplify it.

kanta1207 replied:
At the end, I removed type because it always has to be "search".
But for Icon, there's no negative aspect on keeping this open for overriding, so I'll leave it.

kanta1207 replied:
Fixed this `Clicking somewhere close to the edge does not focus on the input.`.
- Changed input component to be client component, and utilize useRef(). Clicking parent div tag will set focus on input.
- Add "cursor-text" to parent div component to make it more look like input and clickable.

uskayyyyy replied:
Thank you for the update.
It looks like it is working perfectly. But I still have a concern...

We will likely give a `ring` style to each input field shortly and add the `focus:ring-2 ring-primary-dark` classes to the "Input" tag. Using `useRef()`, however, cause an unwanted style like the image below.

<img width="200" alt="Screenshot 2023-12-17 at 16 00 37" src="https://github.com/genesis-tech-tribe/nishiki-frontend/assets/99148565/c533c581-b7dc-4418-a9bc-197f3019c3cc">
<br>
Please remember that when using `<input>` or whatever like that, they should take up the same size as its visual size and shape to avoid unexpected behaviors.

kanta1207 replied:
Indeed, I was blind about that. 
 I'll try to fix it quickly!

kanta1207 replied:
I realized now I forgot to add focus ring to input component in the first place.
In design it's defined as primary-dark, so better to have that style in component, at least for `variant=rounded`

uskayyyyy replied:
@kanta1207 I think you can separate it into a new issue. it was maybe added yesterday. I'll leave the choice to you.

kanta1207 replied:
I fixed the focus issue in SearchInput component! 
I think the component itself got much better now. I should've done it this way in the first place...

Just in case, if it's a bit unclear, pl-12 is `6(icon's left padding) + 4(icon width) + 2(space between icon and text)`!

kanta1207 replied:
Also, "muted" variant is now completely useless, so I removed it at the end 😅

uskayyyyy reviewed (changes requested):
Thank you for implementing it! Please take a look at the comments I left.
And I'd like to have a conversation about the `NumberInput` component.

uskayyyyy commented on src/components/ui/Input/NumberInput.tsx:
You might be able to replace the `onKeyDown` event with `onChange` with `useState`, which might be simpler.
```ts
  const handleChange = (e) => {
    const value = e.target.value.replace(/[^0-9.]/g, '');
    setValue(value);
  };
```

I think these increment/decrement arrows are unnecessary.

<img width="221" alt="image" src="https://github.com/genesis-tech-tribe/nishiki-frontend/assets/99148565/e0137599-2b20-4853-983c-681b5d55f79b">

kanta1207 replied:
>>You might be able to replace the onKeyDown event with onChange with useState, which might be simpler and good enough.

Thank you for suggesting that! 
Increment/decrement arrow is indeed unnecessary, I'll remove it!

And I summarised my thought on your other points, please take a look and let me hear your point of view 🙏 

### Use of "onKeyDown" over "onChange"
I thought we might possibly want to use "onChange" when we use this number input component, maybe for validation or other things. 
For example if user pass "onChange" function to this component like in this screenshot, the "onChange" method inside of this component will be overridden. I wanted to avoid that, so I used "onKeyDown", which seems to be less likely to be used.
<img width="255" alt="Screenshot 2023-12-16 at 21 58 28" src="https://github.com/genesis-tech-tribe/nishiki-frontend/assets/99339182/03c259ce-1072-40bd-b08b-72c044bb335f">

### Not using "useState"
UseState can help us to make this code look simpler, but I thought if it's possible, it would be better not to have unnecessary state. So I chose not to use useState, and used `e.preventDefault()`

### e.target.value.replace(/[^0-9.]/g, '');
It's indeed much more simple, but I wanted to have flexibility on which type of input to decline.
 Also, We anyway need another logic to prevent leading 0 input (numbers which starts with 0)

uskayyyyy replied:
> I thought we might possibly want to use "onChange" when we use this number input component, maybe for validation or other things. For example if user pass "onChange" function to this component like in this screenshot, the "onChange" method inside of this component will be overridden. I wanted to avoid that, so I used "onKeyDown", which seems to be less likely to be used.

I really like this perspective!

You could accept a regular expression as a prop, and utilize it in the `onKeyDown` event handler. I will leave the choice to you.

```ts
  const handleKeyDown = (event) => {
    // Prevent any key press that is not a number or a dot
    if (!/[0-9.]/.test(event.key)) {
      event.preventDefault();
    }
  };
```

uskayyyyy replied:
Don't you think setting the `exceptDot` default to false is better? Because it's more likely to be used with decimals, especially in quantity, which could be "g", "kg", or whatever.

kanta1207 replied:
Yeah, indeed exceptDot can be default to be false! I agree. 
I'll follow your idea on this, thanks 🙌

kotaaaa replied:
I am concerning that if we should make validation logic common. To be honest, I agree with the opinion with using `useState`. but currently I cannot come up with alternative so we can let it go now.

kanta1207 commented on src/components/ui/Input/NumberInput.tsx:
```diff
@@ -0,0 +1,84 @@
+  if (exceptPlusOperator && key === plus) {
+    event.preventDefault();
+  }
+};
```
So sorry this is indeed soooo WET, I was so blind 😅 
I'll fix it!!

kanta1207 commented on src/components/ui/Input/Input.tsx:
```diff
@@ -0,0 +1,48 @@
+  'flex w-full h-full text-base focus:outline-none disabled:cursor-not-allowed',
+  {
+    variants: {
+      variant: {
```
I was thinking of setting the height flexible for user by setting it "h-full".
We can control the size by adding padding in the parent component. 
One of the reason I did that is Because we're not sure yet how much height it will be in desktop design or other designs.

But I'm not sure which is the best practice.

uskayyyyy replied:
I guess that if its parent's paddings control the input size, the input cannot be focused by clicking those paddings, just like what's happening right now in the `SerchInput` component.

uskayyyyy replied:
I think giving `h-full` to the input is an acceptable approach, but in that case, its parent element must have a fixed height without adding paddings.

kanta1207 replied:
I got it!
Either one of input or parent component has to have fixed height, and other UI components(button, labels) already have fixed size. So maybe it's better to have fixed height in UI components. I'll follow your opinion, thank you!!!

kanta1207 replied:
`Please consider adding px to the muted and square variants.`


I've included px or the square variants. 
Regarding the muted variant, I believe zero padding is crucial to keep it useful, controllable from parent component.

kotaaaa commented:
@kanta1207 
Thank you for your implementation!! I have nothing to add other than @uskayyyyy 's points.
After handling points by @uskayyyyy,  you can merge it.

@uskayyyyy 
Thank you for your detailed review!

kanta1207 commented:
@kotaaaa 
Thank you for your comment !

@uskayyyyy 
Thank you for your review! I fixed all the things we discussed about. I added 🚀 stamp to the topic I resolved.
Please take a look at them again on your convenience !

uskayyyyy commented on src/components/ui/Input/InputUsageExample.tsx:
```diff
@@ -0,0 +1,21 @@
+      <Input variant="square" type="text" id="groupName" />
+      <SearchInput placeholder="Search Foods" />
+      <div className="w-1/2">
+        <Label>aaaa</Label>
```
Please add `htmlFor`

uskayyyyy commented on src/components/ui/Input/NumberInput.tsx:
```diff
@@ -0,0 +1,74 @@
+  exceptE = true,
+  exceptLeadingZero = true,
+  exceptNegative = true,
+  exceptPlusOperator = true,
```
```ts
const SYMBOLS = {
  zero: '0',
  negative: '-',
  plus: '+',
  e: ['e', 'E'],
  dot: '.',
};
```

Just like this: 
This collection of similar props can be modularized like:

```ts
export const NumberInput: FC<INumberInputProps> = ({
  className,
  variant = defaultVariant,
  onKeyDown = handleExceptionSymbols,
  options: {
    exceptDot = false,
    exceptE = true,
    exceptLeadingZero = true,
    exceptNegative = true,
    exceptPlusOperator = true, 
  },
  ...props
}) => {

kanta1207 commented:
@uskayyyyy Resolved 2 additional points you mentioned! 
Please confirm them again when you're available!

kanta1207 commented:
Added "h" prop to Input comp, to set the fixed height.  For now there're "h-10" and "h-12" in design, so set them as "sm" and "md". 
The reason why I didn't use "height" was because "height" is already used in default "InputHTMLAttributes", that require us to omit default "height" if we want to use our original "height" property. 
I thought it's better to keep  default "height" property from  "InputHTMLAttributes" as it is, so decided to use "h" as our original height property.

uskayyyyy commented on src/components/ui/Input/Input.tsx:
```diff
@@ -0,0 +1,51 @@
+      muted: 'bg-transparent border-none placeholder:text-gray ',
+      square: 'bg-gray-lightest border-b border-primary p-2',
+    },
+    h: {
```
I suppose naming it `height` would be better than `h` here. Here is my understanding:

When naming something, especially a variable, we should be aware of where it will be used. If its scope is small, a shorter name is usually preferred. If its scope is broad, on the other hand, the name should be easy to understand to guess what it is.

This is an example of the small scope. `f` is used only inside the `forEach`, so it's still understandable.
```ts
const foods = ['onion', 'tomato', 'meat'];
foods.forEach((f) => {
  console.log(f);
})
```

However, this `Input` component would be used outside this file, meaning it has a broader scope. We cannot tell what `h` is until we visit the file it's defined.
```
<Input variant="rounded" h="sm" />
```

uskayyyyy replied:
Where is the height `sm` used, by the way? I couldn't find it in Figma. We would like to avoid defining styles in code that are not defined in Figma.

kanta1207 replied:
I got your point.
Just in case you missed it, my concern on this was https://github.com/genesis-tech-tribe/nishiki-frontend/pull/104#issuecomment-1859338086, but as my understanding you prefer overriding it with our original "height" property, right?

Height "sm" is now used in this square input. 
<img width="414" alt="Screenshot 2023-12-17 at 16 52 17" src="https://github.com/genesis-tech-tribe/nishiki-frontend/assets/99339182/57c7a8d1-6aa0-4ec8-9fb2-3dc82995bc29">

And I suppose in desktop design, there will be larger heights, so we'll add them in "height" property.

uskayyyyy replied:
Ohhh no, I completely missed your comment.

uskayyyyy replied:
> Height "sm" is now used in this square input.

Then, I suppose it's good to have two variants! Thank you!

kanta1207 replied:
Ummm I'm actually changing my mind... Sorry 😅 
Overriding default "InputHTMLAttributes" is maybe inappropriate but not a big problem...
Because we're not gonna use it to be honest.

@uskayyyyy @kotaaaa 
To be honest, do you prefer using simple "height" property? It's indeed much more easy to understand.

uskayyyyy replied:
I do not prefer overriding to avoid unexpected behaviors.

kanta1207 replied:
It actually have "size" property also

uskayyyyy replied:
I tried overriding but it didn't work

kanta1207 replied:
Not sure but we're gonna use input type="image" maybe? To set the profile image of user, or container or something.
So maybe better to keep default height property, and use "h" for setting height?

kanta1207 replied:
Thank your for always considering and testing precisely!!

I think I'll go for `h` for now!

uskayyyyy replied:
When we implement `type="image"`, we would create a new `<ImageImput>` component.
But as you say, I prefer not to override the default attribute.

uskayyyyy reviewed (approved):
Thank you for modifying the code many times. The search input looks perfect to me!
I approve because it's good enough to merge!

As for the input focus style, may I hear why you chose `border` rather than `ring`? After a concern, you can choose either update or merge right away. Thank you for your excellent job!

Small differences I found:

Using `border`, the content shrinks:

https://github.com/genesis-tech-tribe/nishiki-frontend/assets/99148565/d65a7877-066c-4efe-bef4-f53df04ddc06

Using `ring` (or `box-shadow`), the content does not shrink, but the border remains:

https://github.com/genesis-tech-tribe/nishiki-frontend/assets/99148565/d7d668e9-9718-4769-b16c-5248fbe02c8a

Using `ring` with `border-transparent`, the content does not shrink, and the border becomes invisible:

https://github.com/genesis-tech-tribe/nishiki-frontend/assets/99148565/50a83d2f-08fc-4891-bb48-cf52dddc8871

kanta1207 commented:
> Thank you for modifying the code many times. The search input looks perfect to me! I approve because it's good enough to merge!
> 
> As for the input focus style, may I hear why you chose `border` rather than `ring`? After a concern, you can choose either update or merge right away. Thank you for your excellent job!
> 
> Small differences I found:
> 
> Using `border`, the content shrinks: https://github.com/genesis-tech-tribe/nishiki-frontend/assets/99148565/d65a7877-066c-4efe-bef4-f53df04ddc06
> 
> Using `ring` (or `box-shadow`), the content does not shrink, but the border remains: https://github.com/genesis-tech-tribe/nishiki-frontend/assets/99148565/d7d668e9-9718-4769-b16c-5248fbe02c8a
> 
> Using `ring` with `border-transparent`, the content does not shrink, and the border becomes invisible: https://github.com/genesis-tech-tribe/nishiki-frontend/assets/99148565/50a83d2f-08fc-4891-bb48-cf52dddc8871

Thank you for your comment and the precise reviews!
And sorry, should've mentioned that...
The context is, I tried using only "ring", and border persists, so I just went for changing "border".

 Using `ring` with `border-transparent` is definetly the best!! Couldn't come up with that.
 Thank you for bring it up. I will follow your advice and fix it, and then will merge it!
